Objective-C/CBLDatabase.mm

```diff
@@ -425,94 +448,110 @@ - (BOOL) createIndexOn: (NSArray*)expressions
+        if (!json)
+            return NO;
+        C4Error c4err;
+        return c4db_deleteIndex(_c4db, {json.bytes, json.length}, (C4IndexType)type, &c4err)
+        || convertError(c4err, outError);
+    }
 }
 
 
 #pragma mark - INTERNAL
 
+- (C4Database*) c4db {
+    CBL_LOCK(_lock) {
+        return _c4db;
+    }
+}
+
 - (void) postDatabaseChanged {
-    if (!_obs || !_c4db || c4db_isInTransaction(_c4db))
-        return;
-    
-    const uint32_t kMaxChanges = 100u;
-    C4DatabaseChange changes[kMaxChanges];
-    bool external = false;
-    uint32_t nChanges = 0u;
-    NSMutableArray* docIDs = [NSMutableArray new];
-    do {
-        // Read changes in batches of kMaxChanges:
-        bool newExternal;
-        nChanges = c4dbobs_getChanges(_obs, changes, kMaxChanges, &newExternal);
-        if (nChanges == 0 || external != newExternal || docIDs.count > 1000) {
-            if(docIDs.count > 0) {
-                CBLDatabaseChange* change = [[CBLDatabaseChange alloc] initWithDocumentIDs: docIDs
-                                                                                isExternal: external];
-                // Only notify if there are actually changes to send
-                NSDictionary *userInfo = @{ kCBLDatabaseChangesUserInfoKey: change };
-                [[NSNotificationCenter defaultCenter] postNotificationName: kCBLDatabaseChangeNotification
-                                                                    object: self
-                                                                  userInfo: userInfo];
-                docIDs = [NSMutableArray new];
-            }
-        }
+    CBL_LOCK(_lock) {
+        if (!_obs || !_c4db || c4db_isInTransaction(_c4db))
+            return;
         
-        external = newExternal;
-        for(uint32_t i = 0; i < nChanges; i++) {
-            NSString *docID =slice2string(changes[i].docID);
-            [docIDs addObject: docID];
-        }
-    } while(nChanges > 0);
+        const uint32_t kMaxChanges = 100u;
+        C4DatabaseChange changes[kMaxChanges];
+        bool external = false;
+        uint32_t nChanges = 0u;
+        NSMutableArray* docIDs = [NSMutableArray new];
+        do {
+            // Read changes in batches of kMaxChanges:
+            bool newExternal;
+            nChanges = c4dbobs_getChanges(_obs, changes, kMaxChanges, &newExternal);
+            if (nChanges == 0 || external != newExternal || docIDs.count > 1000) {
+                if(docIDs.count > 0) {
+                    CBLDatabaseChange* change = [[CBLDatabaseChange alloc] initWithDocumentIDs: docIDs
+                                                                                    isExternal: external];
+                    NSDictionary *userInfo = @{ kCBLDatabaseChangesUserInfoKey: change };
+                    [[NSNotificationCenter defaultCenter] postNotificationName: kCBLDatabaseChangeNotification
```

Comment:
I agree. Need to think how to avoid this when the notification happens in batch. Probably send the notification async?